### PR TITLE
HPCC-26858 ecl-queries-files improvements

### PR DESCRIPTION
**Pull request by afishbeck**
Precache the data for ecl-queries-files command to greatly improve
performance.

Also add a new option to show the files for the entire target (queryset).

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@afishbeck I think they are not related to your changes. Test rescheduled.

File: ecl/eclcmd/queries/ecl-queries.cpp [hunk 1]
```diff
@@ -368,34 +398,16 @@ class EclCmdQueryFiles : public EclCmdCommon
...
                 {
-                    IConstQuerySuperFile &superfile = superfiles.item(sp);
-                    fprintf(stdout, "  %s\n", superfile.getName());
-                    StringArray &subfiles = superfile.getSubFiles();
-                    ForEachItemIn(sb, subfiles)
-                        fprintf(stdout, "    > %s\n", subfiles.item(sb));
+                    IConstQueryFilesUsed &query = resp->getQueries().item(i);
+                    outputQueryFiles(query.getQueryId(), query.getFiles(), query.getSuperFiles());
                 }
-                fputs("\n", stdout);
```
Comment (wangkx) [inline ecl/eclcmd/queries/ecl-queries.cpp:398]: question: is it possible that the optQuery is not set and the resp->getQueries() returns empty? If yes, should we output a message similar to "No files used.\n" at that situation?

File: esp/services/ws_workunits/ws_workunitsQuerySets.cpp [hunk 2]
```diff
@@ -1895,8 +1875,74 @@ bool CWsWorkunitsEx::onWUQueryFiles(IEspContext &context, IEspWUQueryFilesReques
...
+        queryid.set(registeredQuery->queryProp("@id"));
+    }
+
+    Owned<IPropertyTree> tree = filesInUse.getTree();
+    if (queryid.length())
+    {
+        VStringBuffer xpath("%s/Query[@id='%s']", target, queryid.get());
+        IPropertyTree *queryTree = tree->queryPropTree(xpath);
+        if (!queryTree)
+        throw MakeStringException(ECLWATCH_QUERYID_NOT_FOUND, "Query not found in file cache (%s)", xpath.str());
```
Comment (wangkx) [inline esp/services/ws_workunits/ws_workunitsQuerySets.cpp:1927]: indent here?

File: ecl/eclcmd/queries/ecl-queries.cpp [hunk 3]
```diff
@@ -368,34 +398,16 @@ class EclCmdQueryFiles : public EclCmdCommon
         if (ret == 0)
         {
             IArrayOf<IConstFileUsedByQuery> &files = resp->getFiles();
-            if (!files.length())
-                fputs("No files used.\n", stdout);
+            IArrayOf<IConstQueryFilesUsed> &queries = resp->getQueries();
```
Comment (wangkx) [inline ecl/eclcmd/queries/ecl-queries.cpp:401]: question: how about moving this line before line 406 and use the queries in line 406 and 408? Or remove this line?

**Review by wangkx — COMMENTED** [review]
@afishbeck 3 questions

**Review by ghalliday — COMMENTED** [review]
@afishbeck looks good - a couple of minor comments.

A couple of other comments...  It would be useful to have a comment on getTree() inidcating it is thread safe because it creates a new tree if the tree had changed.  (I double checked because it wasn't obvious).

Also any ideas why the operation is so slow?  @jakesmith is the following line in referencedfilelist.cpp(685) inefficient?:

    Owned<IPropertyTreeIterator> iter = xgmml->getElements("//node[att/@name='_*ileName']");

would it be better to remove the filter (since it is checked in the loop, how efficient is the wildcarding) - although that will significanly increase the side of the target array.  Is it better to iterate each level in turn (I doubt it)?

@afishbeck looking at the code in ReferencedFileList::addFilesFromQuery, it looks like the code will still do lots of work - even if the file has already been resolved from another activity in the same package/  I suspect this is a fairly common occurrence.

**Comment by jakesmith** [comment]
> 
> Also any ideas why the operation is so slow? @jakesmith is the following line in referencedfilelist.cpp(685) inefficient?:
> 
> ```
> Owned<IPropertyTreeIterator> iter = xgmml->getElements("//node[att/@name='_*ileName']");
> ```
> 
> would it be better to remove the filter (since it is checked in the loop, how efficient is the wildcarding) - although that will significanly increase the side of the target array. Is it better to iterate each level in turn (I doubt it)?
> 

not sure how slow it is.
One observation, it looks like getGraphs tries to pre-load/cache the graphs, but doesn't fully load them, meaning they will be fetched from SDS layer by layer on demand. If so, it could help if a queryBranch() was used to ensure the whole graph was loaded in one go.

The wildcards will be preventing the fast cached qualifier maps being used, so linear scans for matches of all nodes at each level, searching for a match will occur (as opposed to single HT lookup at each level).
So I do wonder whether 2 passes, 1 searching for '_fileName' matches, followed by a 2nd searching for '_indexFileName' might be faster.

Is this actually the area of slowness vs the code using the results of this iterator?

**Comment by ghalliday** [comment]
@jakesmith I don't actually know what is causing this code to be slow - I was just reviewing the code to see if anything jumped out.  It would need testing and ideally profiling to find the cause.
I don't think your qualifier maps will help here (will it?) - because the filter is on existence of children.

**Comment by jakesmith** [comment]
> I don't think your qualifier maps will help here (will it?) - because the filter is on existence of children.

you're right, won't apply in this case, only for attribute=value cases.

File: ecl/eclcmd/queries/ecl-queries.cpp (hunk 1 above)
Comment (afishbeck) [reply ecl/eclcmd/queries/ecl-queries.cpp:398]: Sure, I will add a message.

**Comment by afishbeck** [comment]
@richardkchapman @ghalliday updated.  With changes relevant to this PR.  

Any performance improvements can come later.. but I don't think it's that much slower than I would expect.  The perception of slowness was because it rebuilds the entire cache when dirty, not just for a given query.. and odds are this is often called right after dirtying the cache.

The code doesn't do any DALI lookups until it has a deduped list across the entire query.  There may still be some improvements I can make caching some information across multiple queries.

**Review by wangkx — APPROVED** [review]
The changes look fine to me.

**Review by ghalliday — APPROVED** [review]
Looks good, please squash

**Comment by afishbeck** [comment]
@ghalliday squashed

**Pull request merged**